### PR TITLE
PEM: Make PKCS8 serializing functions aware of OSSL_SERIALIZERs

### LINKED ISSUE #11845: OpenSSL 3.0.0-alpha2 can't serialize ed25519 or ed448 keys to PKCS8
The following code will successfully serialize an ed25519 key (or ed448 key if you swap the NID) as a PKCS8 structure in both PEM and DER encoding on OpenSSL 1.1.1, but errors on 3.0.0-alpha2.

```c
#include <stdio.h>
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/pem.h>
#include <openssl/obj_mac.h>
int main(void) {
    int res = 0;

    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(NID_ED25519, NULL);
    res = EVP_PKEY_keygen_init(ctx);
    if (res != 1) {
        printf("could not init keygen\n");
        return 1;
    }
    EVP_PKEY *pkey;
    res = EVP_PKEY_keygen(ctx, &pkey);
    if (res != 1) {
        printf("could not generate key\n");
        return 2;
    }
    BIO *bio = BIO_new(BIO_s_mem());
    res = PEM_write_bio_PKCS8PrivateKey(bio, pkey, NULL, "", 0, NULL, NULL);
    if (res != 1) {
        printf("failed to serialize to PEM\n");
    }
    res = i2d_PKCS8PrivateKey_bio(bio, pkey, NULL, "", 0, NULL, NULL);
    if (res != 1) {
        printf("failed to serialize to DER\n");
    }

    return 0;
}
```

### PR DESCRIPTION
**Pull request by levitte**
PEM_write_bio_PKCS8PrivateKey(), i2d_PKCS8PrivateKey_bio(),
PEM_write_PKCS8PrivateKey(), and i2d_PKCS8PrivateKey_fp() are affected
by this.

Fixes #11845

### CODE REVIEW

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -767,6 +767,40 @@ static int test_EVP_PKCS82PKEY(void)
 }
 #endif
 
+/* This uses kExampleRSAKeyDER and kExampleRSAKeyPKCS8 to verify encoding */
+static int test_privatekey_to_pkcs8(void)
+{
+    EVP_PKEY *pkey;
```
Comment (mattcaswell) [inline test/evp_extra_test.c:773]: Travis failure here. Set this to NULL to avoid unitialised use on an error path

Comment (levitte) [reply test/evp_extra_test.c:773]: Fixed

**Comment by t8m** [comment]
Close/reopen to re-run Travis.

**Comment by levitte** [comment]
CIs are happen, all that's missing is someone's blessing

**Comment by levitte** [comment]
Ping?

File: crypto/pem/pem_pk8.c [hunk 2]
```diff
@@ -66,49 +67,89 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
+    OSSL_SERIALIZER_CTX *ctx = OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(x, pq);
 
-    if ((p8inf = EVP_PKEY2PKCS8(x)) == NULL) {
-        PEMerr(PEM_F_DO_PK8PKEY, PEM_R_ERROR_CONVERTING_PRIVATE_KEY);
+    if (ctx == NULL)
         return 0;
+
+    /*
+     * If no keystring or callback is set, OpenSSL traditionally uses the
+     * user cb argument as a password string, or if that's NULL, it falls
```
Comment (paulidale) [inline crypto/pem/pem_pk8.c:81]: `s/user/user's/`

File: crypto/pem/pem_pk8.c [hunk 3]
```diff
@@ -66,49 +67,89 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
+    if (OSSL_SERIALIZER_CTX_get_serializer(ctx) != NULL) {
+        ret = 1;
+        if (enc != NULL) {
+            ret = 0;
+            if (OSSL_SERIALIZER_CTX_set_cipher(ctx, EVP_CIPHER_name(enc),
+                                               NULL)) {
+                const unsigned char *ukstr = (const unsigned char *)kstr;
+
+                ret = 1;
+                if (kstr != NULL
```
Comment (paulidale) [inline crypto/pem/pem_pk8.c:102]: _You are lost in a maze of nested if statements._
_>_




Well I seem to be.

File: crypto/pem/pem_pk8.c (hunk 2 above)
Comment (levitte) [reply crypto/pem/pem_pk8.c:81]: Fixed

File: crypto/pem/pem_pk8.c (hunk 3 above)
Comment (levitte) [reply crypto/pem/pem_pk8.c:102]: I added a comment

**Comment by levitte** [comment]
Ping?

File: crypto/pem/pem_pk8.c [hunk 4]
```diff
@@ -66,49 +67,95 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
-        if (isder)
-            ret = i2d_PKCS8_PRIV_KEY_INFO_bio(bp, p8inf);
-        else
-            ret = PEM_write_bio_PKCS8_PRIV_KEY_INFO(bp, p8inf);
-        PKCS8_PRIV_KEY_INFO_free(p8inf);
-        return ret;
+        X509_SIG *p8;
+        PKCS8_PRIV_KEY_INFO *p8inf;
+        char buf[PEM_BUFSIZE];
+
```
Comment (t8m) [inline crypto/pem/pem_pk8.c:122]: You're leaking the ctx in error cases here in the legacy code.

File: crypto/pem/pem_pk8.c [hunk 5]
```diff
@@ -66,49 +67,95 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
+    if (OSSL_SERIALIZER_CTX_get_serializer(ctx) != NULL) {
+        ret = 1;
+        if (enc != NULL) {
+            ret = 0;
+            if (OSSL_SERIALIZER_CTX_set_cipher(ctx, EVP_CIPHER_name(enc),
+                                               NULL)) {
+                const unsigned char *ukstr = (const unsigned char *)kstr;
+
+                /*
+                 * Try to pass the passphrase if one was given, of the
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/pem/pem_pk8.c:102]: Fixed

File: crypto/pem/pem_pk8.c (hunk 4 above)
Comment (levitte) [reply crypto/pem/pem_pk8.c:122]: Fixed

File: crypto/pem/pem_pk8.c [hunk 6]
```diff
@@ -66,49 +67,94 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
+        if ((p8inf = EVP_PKEY2PKCS8(x)) == NULL) {
+            PEMerr(PEM_F_DO_PK8PKEY, PEM_R_ERROR_CONVERTING_PRIVATE_KEY);
+            ret = 0;
+        } else if (enc || (nid != -1)) {
+            if (kstr == NULL) {
+                klen = cb(buf, PEM_BUFSIZE, 1, u);
+                if (klen <= 0) {
+                    PEMerr(PEM_F_DO_PK8PKEY, PEM_R_READ_KEY);
+                    PKCS8_PRIV_KEY_INFO_free(p8inf);
+                    return 0;
```
Comment (t8m) [inline crypto/pem/pem_pk8.c:132]: There is still a leak here.

File: crypto/pem/pem_pk8.c [hunk 7]
```diff
@@ -66,49 +67,94 @@ static int do_pk8pkey(BIO *bp, const EVP_PKEY *x, int isder, int nid,
...
+                }
+
+                kstr = buf;
+            }
+            p8 = PKCS8_encrypt(nid, enc, kstr, klen, NULL, 0, 0, p8inf);
+            if (kstr == buf)
+                OPENSSL_cleanse(buf, klen);
+            PKCS8_PRIV_KEY_INFO_free(p8inf);
+            if (p8 == NULL)
+                return 0;
```
Comment (t8m) [inline crypto/pem/pem_pk8.c:142]: And here. Sorry for not making that clear before.

File: crypto/pem/pem_pk8.c (hunk 6 above)
Comment (levitte) [reply crypto/pem/pem_pk8.c]: Darnit...

File: crypto/pem/pem_pk8.c (hunk 7 above)
Comment (levitte) [reply crypto/pem/pem_pk8.c]: No problem, my own fault...  darnit.

Fixed, I hope.

**Comment by levitte** [comment]
Ping

**Comment by levitte** [comment]
Merged

eb2bba2569a3637b8477c56f82b616222f5af46f PEM: Make PKCS8 serializers aware of OSSL_SERIALIZERs
29c49b2534fbd60338f61e94c2893d774f9361a9 TEST: Test i2d_PKCS8PrivateKey_bio() and PEM_write_bio_PKCS8PrivateKey()

**Pull request closed without merging**